### ISSUE TITLE
Neo v3.7.0 plan

### ISSUE THREAD

**Issue by roman-khimov** [neo-project/neo#2905]
**Summary or problem description**
3.6.0 is out for a week. And we don't have any roadmap for 3.7. Fail to plan is plan to fail, so there has to be some.

**Do you have any solution you want to propose?**
We have a lot of open issues and PRs. In general my approach to them is:
 * simple and short problems should not sit in the queue for years, they're either accepted or closed (SJF)
 * anything affecting already existing applications has a priority (especially, bugs)
 * protocol changes should have some impact, enabling (opening new possibilities for a wide variety of applications) technologies are prioritized
 * sitting on a bunch of changes for a long time doesn't help, release more often

**The list**

Issues from this repository were moved into milestone: https://github.com/neo-project/neo/milestone/2
Below other repositories are being tracked along with items that require additional discussion.

 * https://github.com/neo-project/neo-modules/issues/814
 * https://github.com/neo-project/neo-modules/issues/811
 * https://github.com/neo-project/neo-node/pull/905
 * https://github.com/neo-project/neo-modules/issues/852
 * https://github.com/neo-project/neo-modules/issues/877
 * https://github.com/neo-project/neo-modules/issues/850
 * https://github.com/neo-project/neo-modules/issues/854
 * https://github.com/neo-project/neo-modules/pull/859
 * https://github.com/neo-project/neo-modules/pull/863
 * https://github.com/neo-project/neo-modules/pull/845
 * https://github.com/neo-project/neo-modules/issues/897
 * https://github.com/neo-project/neo-modules/pull/896

Potentially 3.8:
 * https://github.com/neo-project/proposals/pull/160
 * https://github.com/neo-project/neo-modules/pull/865

**The date**
Dec 15 seems to be a good one for the list, enough to solve most of the important problems and with some room for unexpected things. The list can be adjusted if there are any problems, but there has to be a nice release this year .

**Discussion**
@superboyiii, @steven1227, core developers, Neo developers, dApp developers, random guys from the internet --- feel free to drop/replace this issue or edit it in any way. I'm not pretending that this is _the_ list everyone should follow. Propose alternatives, kick some things out and bring interesting stuff in. The main thing here is to get this thing going and have at least some rough roadmap and schedule.

**Comment by superboyiii** [comment]
@neo-project/core @neo-project/ngd-seattle @neo-project/developers  We could make discussion about this checklist as early as possible, especially for notary related proposals.

**Comment by cschuchardt88** [comment]
Add https://github.com/neo-project/neo-modules/pull/823

**Comment by roman-khimov** [comment]
It's there as neo-project/neo-modules#822 (one entry is enough I think). Likely we need https://github.com/neo-project/neo-modules/pull/825 though.

**Comment by vncoelho** [comment]
@roman-khimov, I was checking the proposed mandatory PRs.
Maybe just merge all three notary on the list into one (create an issue that list all three of them) because they need to be applied together.

Maybe postpone https://github.com/neo-project/proposals/pull/160 because of time. 
I think that Notary will take some time and the others PRs you defined as mandatory are kind of important ( I believe they were a good pick).

**Comment by vncoelho** [comment]
I understand, @roman-khimov, For https://github.com/neo-project/neo/issues/2895 I do not see a reason to add role if the functionality is not implemented. I think it can be merged to another issue related to Notary.

Let's see how the extension goes in practice. We also need to study it careful to see possible impacts. It is also early as well, we reevaluate this later.

**Comment by cschuchardt88** [comment]
`Neo` really needs to make the system stable. Before adding new features to the `neo` ecosystem. It very annoying to develop on `neo` where it takes a lifetime to get a bug fix. These releases come out to **FAST**!!! If `neo` took the time to develop and actually a dedicated QA Test team. We wouldn't be rushing for new version releases as soon as a release comes out.

So i say 
- https://github.com/neo-project/neo-modules/issues/814
- https://github.com/neo-project/neo/issues/2895
- https://github.com/neo-project/neo/issues/2896
- https://github.com/neo-project/neo/issues/2896
- https://github.com/neo-project/proposals/pull/160

should all wait until 4.0

**Comment by roman-khimov** [comment]
I've removed the ones included in 3.6.2, some other updates will be proposed later. I'd use GH milestones and leave this issue for general discussion (which will be easier to do after #2944, now we still have references to other repositories).

**Comment by roman-khimov** [comment]
It got some traction already, so likely yes. My initial thought was to limit it to vm/node part of the deal, but maybe we could arrange a complete solution (well, it's just modules besides that two) in 3.7.0 time frame. Not release critical to me, though. As usual, I'm a bit more concerned about _protocol_ changes or fixes we deliver.

**Comment by roman-khimov** [comment]
So, I've created https://github.com/neo-project/neo/milestone/2, #2974 and moved some things there. I still think we should have a release this year and with the current set of issues there are some chances for it (+/- some items). 3.7 will require a resync and I don't believe we can handle #1526 quickly, so it's moved to 3.8.

**Comment by lock9** [comment]
>@roman-khimov, should we include https://github.com/neo-project/neo/issues/2944 into 3.7.0?

Maybe we could add #2976 instead of #2944?

**Comment by roman-khimov** [comment]
There are three things to merge (vm/node/modules). Go one by one. It doesn't really matter how many will be included into 3.7.0, from the user perspective it doesn't change the resulting product at all (at least until we make modules built-in, but that's not a quick thing even if we'd like to).

**Comment by shargon** [comment]
#2944 and #2942

**Comment by Liaojinghui** [comment]
> There are three things to merge (vm/node/modules). Go one by one. It doesn't really matter how many will be included into 3.7.0, from the user perspective it doesn't change the resulting product at all (at least until we make modules built-in, but that's not a quick thing even if we'd like to).

I am seriously suggest adding compiler as well, debugging the compiler is also painful and need references to the neo and neo-vm.

**Comment by shargon** [comment]
> > There are three things to merge (vm/node/modules). Go one by one. It doesn't really matter how many will be included into 3.7.0, from the user perspective it doesn't change the resulting product at all (at least until we make modules built-in, but that's not a quick thing even if we'd like to).
> 
> I am seriously suggest adding compiler as well, debugging the compiler is also painful and need references to the neo and neo-vm.

If we override some methods we can make it easier

**Comment by AnnaShaleva** [comment]
@roman-khimov, we should include https://github.com/neo-project/neo-modules/issues/850.

And I'd suggest to consider adding https://github.com/neo-project/neo-modules/issues/852 (or at least discussing it). It's a minor issue, but it's still an issue.

**Comment by gsmachado** [comment]
Hey @shargon @Jim8y @roman-khimov 
What's the timeline for v3.7.0? (if any)

**Comment by roman-khimov** [comment]
I'd love for it to happen end of March, but it'll released when it's done and realistically it's more like April to me. IIUC devpack already receives a lot of improvements independent of core changes and we need to solve a number of problems in core before release.

**Comment by vncoelho** [comment]
Can you update the checklist @roman-khimov ?

Maybe it would be better to release a version without notory and then release 3.8.0 quicker with this feature.

**Comment by roman-khimov** [comment]
3.7.0 will inevitably require a resync. Let us hope that it's the last one in our history. For that to be true notary _contract_ needs to be integrated into it. I hope we'll have some time next week to start dealing with that.

As for the list, I know we have a bit more things merged now, but still these are the target ones (see https://github.com/neo-project/neo/milestone/2 also) and things we have merged mostly don't affect the protocol. The only thing that can be moved to 3.8 is https://github.com/neo-project/proposals/pull/160 probably, it'll require some time and at least it doesn't affect state in any way.

**Comment by shargon** [comment]
I think that notary can wait, we need a release before that

**Comment by vncoelho** [comment]
In fact I think the same,@shargon 

We can release another as soon as notary is ready

**Comment by roman-khimov** [comment]
> I think that notary can wait, we need a release before that

Then 3.8 will require a resync.

**Comment by roman-khimov** [comment]
I've updated the milestone with things that affect our deliverables, an impressive list even though protocol-wise it's mostly compatible.

**Comment by roman-khimov** [comment]
BTW, regarding state compatibility. Technically, https://github.com/neo-project/neo-modules/pull/865 is similar here, even though it's a plugin it's a very very popular one and this update (if we're interested in it, it still mixes too many things at once to me) requires plugin DB resynchronization which to users is the same as full resync.

**Comment by shargon** [comment]
> > I think that notary can wait, we need a release before that
> 
> Then 3.8 will require a resync.

Why? now we have upgradeable native contracts

**Comment by shargon** [comment]
> BTW, regarding state compatibility. Technically, [neo-project/neo-modules#865](https://github.com/neo-project/neo-modules/pull/865) is similar here, even though it's a plugin it's a very very popular one and this update (if we're interested in it, it still mixes too many things at once to me) requires plugin DB resynchronization which to users is the same as full resync.

I think we sould not wait for a plugin to release a version

**Comment by Jim8y** [comment]
I have no special idea on if we should do something. But I hope this release is the last release that we need user to resync whole chain.

**Comment by roman-khimov** [comment]
> now we have upgradeable native contracts

Yeah, but we want C# node to be able to process NeoFS chains (of which there are two). For that to happen notary contract needs to be present since the genesis.

**Comment by gsmachado** [comment]
> I have no special idea on if we should do something. But I hope this release is the last release that we need user to resync whole chain. (but not likely)

God bless you, and all Neo users, to endure one more resync. ⚰️ 

I'll cry with happiness the day we roll out features without a full re-sync. 😭 🎉 

Keep up the great work to make it happen. 🚀

**Comment by roman-khimov** [comment]
> I think we should not wait for a plugin to release a version

While idealistically that's the case, reality is --- users won't care much. This plugin (along with RPC server) is used by everyone and their dog, it's kinda a part of a standard installation (if it'd be something like `StatesDumper` that'd be a different story). So if you say "3.8 core doesn't require resync, but this plugin does" for users it still means "ugh, 3.8 requires a resync _again_".

**Comment by shargon** [comment]
> > now we have upgradeable native contracts
> 
> Yeah, but we want C# node to be able to process NeoFS chains (of which there are two). For that to happen notary contract needs to be present since the genesis.

Why from genesis?

**Comment by shargon** [comment]
> Transactions will fail otherwise.

But transactions will be executed with the contract already deployed, I'm missing something?

**Comment by AnnaShaleva** [comment]
Should include https://github.com/neo-project/neo-modules/issues/897 into release plan.

**Comment by AnnaShaleva** [comment]
@roman-khimov, we should include https://github.com/neo-project/neo/issues/3194 into the plan as far.

**Comment by mialbu** [comment]
Hey guys, what's the state of this? Is there an ETA for Neo 3.7.0? Just that we can plan a bit.

@roman-khimov @AnnaShaleva @shargon @Jim8y

**Comment by roman-khimov** [comment]
Seems like it'll happen when it's done. We've got an urgent request in #3205, it needs to be addressed (see https://github.com/nspcc-dev/neo-go/pull/3425 also). Then there is a blocker in https://github.com/neo-project/neo-modules/issues/897. Everything else seems to be optional even though we strongly suggest #2896 and #2897 (both have associated PRs and delaying them complicates future network maintenance).

**Comment by cschuchardt88** [comment]
@roman-khimov
https://github.com/neo-project/neo-modules/issues/897 with PR https://github.com/neo-project/neo-modules/pull/898 should be ready to go. Been waiting for testing and review to happen. Seems last two weeks no one around. Everyone on vacation?

**Comment by AnnaShaleva** [comment]
@roman-khimov, we need to include https://github.com/neo-project/neo/issues/2805 into the plan, otherwise users won't be able to properly calculate network fee for #3209 with C# node. It works OK with Go node, so WIP on this issue for C#.

**Comment by roman-khimov** [comment]
I think we can handle #3209 without #2805 here. #2805 is much better, but it's invasive. Too risky for 3.7 to me, I'd try minimizing changes. Now that K-scripts are easy enough to parse we can have old-fashioned fee calculation scheme for them.

**Comment by AnnaShaleva** [comment]
@roman-khimov, we need to include https://github.com/neo-project/neo/pull/3211, it's needed for proper work of #3209.

**Comment by AnnaShaleva** [comment]
@roman-khimov, we also need to include https://github.com/neo-project/neo/issues/3200 into 3.7. There's already a PR that closes this issue, and it's LGTM.

**Comment by cschuchardt88** [comment]
@roman-khimov 

Please include
- #3221

**Comment by dusmart** [comment]
https://github.com/neo-project/neo-modules/issues/811 added twice, pls delete one

**Comment by superboyiii** [comment]
The master branch has been frozen and is ready to release v3.7.0，please don't merge any PR except change logs. @shargon @cschuchardt88 @Jim8y @roman-khimov @AnnaShaleva

**Comment by roman-khimov** [comment]
Updated the list above and https://github.com/neo-project/neo/milestone/2 (at least in things that affect deliverables). 3.7 is out.

**Comment by dusmart** [comment]
congratulations for 3.7.0 coming out, hopefully the last one that needs re-sync

**Comment by roman-khimov** [comment]
> hopefully the last one that needs re-sync

This goal was certainly not reached this time.

**Comment by gsmachado** [comment]
> This goal was certainly not reached this time.

It means that to go from 3.7.0 to 3.7.1 would potentially need re-sync? 😭
